socketio/src/client/reconnect.rs

```diff
@@ -0,0 +1,239 @@
+                // Lock is poisoned, our iterator is useless.
+                None
+            }
+        }
+    }
+}
+
+#[cfg(test)]
+mod test {
+    use std::sync::mpsc;
+    use std::thread::sleep;
+
+    use super::*;
+    use crate::error::Result;
+    use crate::{client::TransportType, payload::Payload, ClientBuilder};
+    use bytes::Bytes;
+    use native_tls::TlsConnector;
+    use serde_json::json;
+    use std::time::Duration;
+
+    #[test]
+    fn socket_io_reconnect_integration() -> Result<()> {
+        let url = crate::test::socket_io_restart_server();
+        let connect_count = Arc::new(Mutex::new(0));
+        let close_count = Arc::new(Mutex::new(0));
+        let connect_count_clone = connect_count.clone();
+        let close_count_clone = close_count.clone();
+
+        let socket = ClientBuilder::new(url)
+            .reconnect(true)
+            .max_reconnect_attempts(0) // infinity
+            .reconnect_delay(100, 100)
+            .on(Event::Connect, move |_, socket| {
+                let mut count = connect_count_clone.lock().unwrap();
+                *count += 1;
+            })
+            .on(Event::Close, move |_, socket| {
+                let mut count = close_count_clone.lock().unwrap();
+                *count += 1;
+            })
+            .connect();
+
+        for _ in 0..10 {
+            std::thread::sleep(std::time::Duration::from_millis(100));
+            {
+                let open_cnt = connect_count.lock().unwrap();
+                if *open_cnt == 1 {
+                    break;
+                }
+            }
+        }
+
+        {
+            let connect_cnt = connect_count.lock().unwrap();
+            let close_cnt = close_count.lock().unwrap();
+            assert_eq!(*connect_cnt, 1, "should connect once ");
+            assert_eq!(*close_cnt, 0, "should not close");
+        }
+
+        assert!(socket.is_ok(), "should connect success");
+        let socket = socket.unwrap();
+
+        let r = socket.emit_with_ack("message", json!(""), Duration::from_millis(100), |_, _| {});
+        assert!(r.is_ok(), "should emit message success");
+
+        let r = socket.emit("restart_server", json!(""));
+        assert!(r.is_ok(), "should emit restart success");
+
+        for _ in 0..10 {
+            std::thread::sleep(std::time::Duration::from_millis(500));
+            {
+                let open_cnt = connect_count.lock().unwrap();
+                if *open_cnt == 2 {
+                    break;
+                }
+            }
+        }
+
+        {
+            let connect_cnt = connect_count.lock().unwrap();
+            let close_cnt = close_count.lock().unwrap();
+            assert_eq!(
+                *connect_cnt, 2,
+                "should connect twice while server is gone and back"
+            );
+            assert_eq!(*close_cnt, 1, "should close once while server is gone");
+        }
+
+        socket.disconnect()?;
+        Ok(())
+    }
```

Comment:
suggestion: test the iterator implementation and make sure there is a constant stream of packets, even when reconnecting

Reply:
how about add 
```rust
.on("message", move |_, socket| {
                // test the iterator implementation and make sure there is a constant
                // stream of packets, even when reconnecting
                let mut message_cnt = message_count_clone.lock().unwrap();
                *message_cnt += 1;
            })
```

Reply:
Prefer if the iterator itself was tested (as the callbacks are meerly another layer on top of the iterator), but that would help a tiny bit